lib/src/lifecycle_module.dart

```diff
@@ -274,36 +366,54 @@ abstract class LifecycleModule extends Object with Disposable {
 
   /// Public method to trigger the Module unload cycle.
   ///
-  /// Calls shouldUnload(), and, if that completes successfully,
-  /// continues to call onUnload() on the module and all registered child modules.
-  /// If unloading is rejected, this method will complete with an error.
+  /// Calls shouldUnload(), and, if that completes successfully, continues to
+  /// call onUnload() on the module and all registered child modules. If
+  /// unloading is rejected, this method will complete with an error.
+  ///
+  /// Initiates the unload process when the module is in the loaded or suspended
+  /// state. If the module is in the unloading or unloaded state a warning is
+  /// logged and the method is a noop. If the module is in any other state, a
+  /// StateError is thrown.
   Future<Null> unload() {
-    final completer = new Completer<Null>();
-    if (_isLoaded) {
-      ShouldUnloadResult canUnload = shouldUnload();
-      if (canUnload.shouldUnload) {
-        _isUnloadedOrUnloading = true;
-        _willUnloadController.add(this);
-        final unloadChildren = _childModules.map((c) => c.unload());
-        Future.wait(unloadChildren).then((_) {
-          _childModules.clear();
-          onUnload().then((_) {
-            _didUnloadController.add(this);
-            _isLoaded = false;
-            _isSuspended = false;
-            super.dispose().then((_) {
-              completer.complete();
-            });
-          });
-        });
-      } else {
-        // reject with shouldUnload messages
-        throw new ModuleUnloadCanceledException(canUnload.messagesAsString());
-      }
-    } else {
+    if (isUnloading) {
+      _logger.warning('Module "$name" is already unloading, cannot unload.');
+      return _transitionFuture;
+    }
+
+    if (isUnloaded) {
       _logger.warning('Module "$name" is already unloaded, cannot unload.');
-      completer.complete();
+      return new Future.value(null);
     }
+
+    if (!(isLoaded || isSuspended)) {
+      return new Future.error(
+          new StateError('Module "$name" state is $_state, cannot unload.'));
+    }
+
+    ShouldUnloadResult canUnload = shouldUnload();
+    if (!canUnload.shouldUnload) {
+      // reject with shouldUnload messages
+      return new Future.error(
+          new ModuleUnloadCanceledException(canUnload.messagesAsString()));
+    }
+
+    _state = LifecycleState.unloading;
+    final completer = new Completer<Null>();
+    _transitionFuture = completer.future;
+    _willUnloadController.add(this);
+
+    final unloadChildren = _childModules.map((c) => c.unload());
+    Future.wait(unloadChildren).then((_) {
+      _childModules.clear();
+      onUnload().then((_) {
+        _didUnloadController.add(this);
+        super.dispose().then((_) {
+          _state = LifecycleState.unloaded;
```

Comment:
This seems odd to emit the `didUnload` event and change the `_state` at different times. If disposal is a part of unloading, I think the `didUnload` event should be emitted when that has finished